### PR TITLE
fix(core): derive possible null value in rxResource stream params (#6…

### LINKED ISSUE #62724: rxResource `params` field type in `stream` function doesn't derive possible null value
### Which @angular/* package(s) are the source of the bug?

core

### Is this a regression?

Yes

### Description

There is a specific case when `(() => Smth) | undefined` is passed to `rxResource.params` field that ends with a wrong type of `rxResource.stream.params` field - possible `null` value is not mentioned in the type.

code (or check the reproduction link)
```ts
params = signal<Params | undefined>(undefined);
data = rxResource({
  params: this.getParams(),
  stream: ({ params }) => {
    console.log('> params', params);
    return NEVER;
  },
});

getParams(): Params | undefined {
  if (!this.params()) {
    return undefined;
  }

  return () => this.params();
}
```

`params` type inside `stream` function is `(parameter) params: Params` but in reality its value can be `null` if `getParams()` returns `undefined`

### Please provide a link to a minimal reproduction of the bug

https://stackblitz.com/edit/kuh9jtu5?file=src%2Fmain.ts

### Please provide the exception or error you saw

```true

```

### Please provide the environment you discovered this bug in (run `ng version`)

```true
Angular CLI: 20.1.1
Node: 24.1.0
Package Manager: npm 11.3.0
OS: darwin arm64

Angular: 20.1.2
... animations, common, compiler, compiler-cli, core, forms
... platform-browser, platform-browser-dynamic, router

Package                      Version
------------------------------------------------------
@angular-devkit/architect    0.2001.1
@angular-devkit/core         20.1.1
@angular-devkit/schematics   20.1.1
@angular/build               20.1.1
@angular/cli                 20.1.1
@schematics/angular          20.1.1
rxjs                         7.8.2
typescript                   5.8.3
```

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by surajy93**
When an optional params function returns `undefined`, Angular's `rxResource` internally defaults the params loader value to `null` at runtime. This change updates `ResourceLoaderParams<R>` to correctly allow `null` alongside the resolved type, ensuring developers get accurate TypeScript feedback when omitted.

Fixes #62724

## PR Checklist

Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/main/contributing-docs/commit-message-guidelines.md
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type

What kind of change does this PR introduce?

- [x] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.dev application / infrastructure changes
- [ ] Other... Please describe:

## What is the current behavior?

When the `params` option is not provided to `rxResource` or it evaluates to `undefined`, the type of the `params` argument in the internal `stream` function is restricted to the strictly inferred type (e.g. `Params`), even though it resolves to `null` at runtime.

Issue Number: #62724

## What is the new behavior?

The type inference for `ResourceLoaderParams<R>` correctly derives a potential `null` value in instances where `undefined` could be inferred from the original `params` option payload. This correctly aligns the TypeScript expectations within `stream` or `loader` functions to match the runtime behavior.

## Does this PR introduce a breaking change?

- [x] Yes
- [ ] No

## Other information
N/A

### CODE REVIEW

File: packages/forms/signals/test/node/resource.spec.ts [hunk 1]
```diff
@@ -69,7 +69,7 @@ describe('resources', () => {
       const RES = createManagedMetadataKey((_state, params: Signal<{x: string} | undefined>) =>
         resource({
           params,
-          loader: async ({params}) => `got: ${params.x}`,
+          loader: async ({params}) => `got: ${params!.x}`,
```
Comment (JeanMeche) [inline packages/forms/signals/test/node/resource.spec.ts]: This is still not an acceptable solution. 
The type shouldn't show up as nullable if the param isn't.

Comment (surajy93) [reply packages/forms/signals/test/node/resource.spec.ts]: thanks i will fix it

Comment (surajy93) [reply packages/forms/signals/test/node/resource.spec.ts]: kindly review again please @JeanMeche

Comment (surajy93) [reply packages/forms/signals/test/node/resource.spec.ts]: don't close the PR checking the test cases why it is falling because locally it is working fine.
if you are closing please provide something so that next time i will make sure of that :)

Comment (JeanMeche) [reply packages/forms/signals/test/node/resource.spec.ts]: AS mentionned in the comments in the issue. The problem also exists on `resource()`.

Comment (surajy93) [reply packages/forms/signals/test/node/resource.spec.ts]: ohh let me check the core resource file too

Comment (surajy93) [reply packages/forms/signals/test/node/resource.spec.ts]: please review again now fixed the resource too @JeanMeche

**Comment by JeanMeche** [comment]
It looks better.
Another example that still doesn't work: 

```
const data = resource({
  params: () => Math.random() ? undefined : 'test',
  loader: ({ params }) => {
    console.log('> params', params); // string, should be string|undefined 
    return Promise.resolve('');
  },
});
```

**Comment by surajy93** [comment]
> It looks better. Another example that still doesn't work:
> 
> ```
> const data = resource({
>   params: () => Math.random() ? undefined : 'test',
>   loader: ({ params }) => {
>     console.log('> params', params); // string, should be string|undefined 
>     return Promise.resolve('');
>   },
> });
> ```



> It looks better. Another example that still doesn't work:
> 
> ```
> const data = resource({
>   params: () => Math.random() ? undefined : 'test',
>   loader: ({ params }) => {
>     console.log('> params', params); // string, should be string|undefined 
>     return Promise.resolve('');
>   },
> });
> ```

let me check  and test it please provide all the test cases and details thank you

**Comment by JeanMeche** [comment]
Also please check the CLA check. You're authoring your commits with 2 different names.

**Comment by surajy93** [comment]
> > ```
> > const data = resource({
> >   params: () => Math.random() ? undefined : 'test',
> >   loader: ({ params }) => {
> >     console.log('> params', params); // string, should be string|undefined 
> >     return Promise.resolve('');
> >   },
> > });
> > ```
> 
> 
> >

@JeanMeche 

loader should stay typed as string here, not string | undefined. When a params function is provided, Angular treats undefined as “no request” and leaves the resource idle, so the loader is not invoked for that branch. In other words, params() may return string | undefined, but loader only runs for the resolved, non-undefined case. The nullable case we’re fixing is different: it applies when the params option itself is omitted or explicitly undefined, because Angular uses an internal null sentinel there.`

Shorter version:

I don’t think this one should widen to \string | undefined`. If `params()` returns `undefined`, the resource goes idle and `loader` is skipped, so inside `loader` the type should remain the non-undefined branch (`string`). The bug we’re fixing is the separate case where the `params` option is omitted or explicitly `undefined`, which currently flows a runtime `null` sentinel.`

please comment or suggestion why do u think that it should be string |undefined?

**Comment by surajy93** [comment]
@JeanMeche apart the this edge cases everything look fine can you please take a look

**Comment by JeanMeche** [comment]
Yeah you're right, my bad for that bad example. 

Something else. 
The generics are somehow not ideal. 

In your test you have: 
```
    const res = resource<string, string>({
      loader: async ({params}) => {
        seenParams = params;
        return 'foo';
      },
      injector: TestBed.inject(Injector),
    });
```

which should arguably have a single generic (since the request is not defined). 

```
    const res = resource<string>({
      loader: async ({params}) => {
        seenParams = params;
        return 'foo';
      },
      injector: TestBed.inject(Injector),
    });
```

**Comment by surajy93** [comment]
> Yeah you're right, my bad for that bad example.
> 
> Something else. The generics are somehow not ideal.
> 
> In your test you have:
> 
> ```
>     const res = resource<string, string>({
>       loader: async ({params}) => {
>         seenParams = params;
>         return 'foo';
>       },
>       injector: TestBed.inject(Injector),
>     });
> ```
> 
> which should arguably have a single generic (since the request is not defined).
> 
> ```
>     const res = resource<string>({
>       loader: async ({params}) => {
>         seenParams = params;
>         return 'foo';
>       },
>       injector: TestBed.inject(Injector),
>     });
> ```

@JeanMeche please review again

**Review by surajy93 — COMMENTED** [review]
kindly review and let me know if any concern

**Review by surajy93 — COMMENTED** [review]
I have fix the test failed can you please again rerun the pipeline @JeanMeche

File: packages/core/src/resource/resource.ts [hunk 2]
```diff
@@ -423,13 +452,20 @@ export class ResourceImpl<T, R> extends BaseWritableResource<T> implements Resou
...
+                abortSignal,
+                previous: {status: previousStatus},
+              }
+            : {
+                params: requestValue,
+                abortSignal,
+                previous: {status: previousStatus},
+              }
+        ) as SafeLoaderParams<R>;
+        return this.loaderFn(loaderParams);
```
Comment (JeanMeche) [inline packages/core/src/resource/resource.ts]: There shouldn't any runtime changes.

**Comment by JeanMeche** [comment]
I would also expect test that show that the typings are correct here.

**Review by surajy93 — COMMENTED** [review]
@JeanMeche i have cover all the test cases and revert the run time code as mention , 
example below : 
  //params ommitted
  resource1 = resource({
    loader: async ({params}) => {
      console.log(params); // {}  | null | undefined
      return 'hello';
    }
  })

  paramsSignal = signal<Params | undefined>(undefined);

  // params are optional
   resource2 = resource({
    params: this.getParams(),
    loader: async ({params}) => {
      console.log(params); // string | undefined
      return 'hello';
    }
  })

  resource3 = rxResource({
    params: this.getParams(),
    stream: ({params}) => {
     console.log(params); // string | undefined
     return of(params ? `id=${params}` : 'no id');
    }
  })

  resource4 = resource({
    params: () => 'test',
    loader: async ({params}) => {
      console.log(params) // string
      return 'hello'
    }
  })

  resource5 = resource({
    params: () => 0,
    loader: async ({params}) => {
      console.log(params) // number
      return 'hello'
    }
  })

   resource6 =rxResource({
    params: () => 'test',
    stream: ({params}) => {
      console.log(params) // string
      return of('hello')
    }
  })

  resource7 =   rxResource({
    params: () => 0,
    stream: ({params}) => {
      console.log(params) // number
      return of('hello')
    }
  })

  getParams(): (() => string) | undefined {
    return Math.random() > 0.5 ? () => 'test' : undefined;
  }

please review again and provide the feedback :)

File: packages/core/src/resource/resource.ts (hunk 2 above)
Comment (surajy93) [reply packages/core/src/resource/resource.ts]: done revert back thanks

File: packages/core/src/resource/api.ts [hunk 3]
```diff
@@ -186,20 +186,30 @@ export interface ResourceLoaderParams<R> {
   };
 }
 
+/**
+ * Safe loader parameters that excludes params when R = never.
+ * When R = never, the loader cannot access params at all.
+ *
+ * @experimental
+ */
+export type SafeLoaderParams<R> = [R] extends [never]
```
Comment (JeanMeche) [inline packages/core/src/resource/api.ts]: It looks to me, this type is not necessary and we can just go with `ResourceLoaderParams` instead.

File: packages/core/rxjs-interop/src/rx_resource.ts [hunk 4]
```diff
@@ -75,16 +122,16 @@ export function rxResource<T, R>(opts: RxResourceOptions<T, R>): ResourceRef<T |
         resolve = undefined;
       }
 
-      const streamFn = opts.stream;
+      const streamFn = opts.stream as (params: SafeLoaderParams<R>) => Observable<T>;
```
Comment (JeanMeche) [inline packages/core/rxjs-interop/src/rx_resource.ts]: Which also means this type assertions isn't necessary

Comment (surajy93) [reply packages/core/rxjs-interop/src/rx_resource.ts]: yes remove

File: packages/core/src/resource/api.ts (hunk 3 above)
Comment (surajy93) [reply packages/core/src/resource/api.ts]: remove and update with ResourceLoaderParams

**Review by surajy93 — COMMENTED** [review]
I have resolve the comment issue mention @JeanMeche  please review again

File: packages/core/src/resource/resource.ts [hunk 5]
```diff
@@ -47,21 +64,32 @@ import {linkedSignal} from '../render3/reactivity/linked_signal';
...
- * Constructs a `Resource` that projects a reactive request to an asynchronous operation defined by
- * a loader function, which exposes the result of the loading operation via signals.
- *
- * Note that `resource` is intended for _read_ operations, not operations which perform mutations.
- * `resource` will cancel in-progress loads via the `AbortSignal` when destroyed or when a new
- * request object becomes available, which could prematurely abort mutations.
- *
- * @experimental 19.0
- * @see [Async reactivity with resources](guide/signals/resource)
- */
```
Comment (SkyZeroZx) [inline packages/core/src/resource/resource.ts:63]: I don't think we should delete the tsDocs file we have (similar to another one we have above)

**Comment by JeanMeche** [comment]
I have squashed & pushed something that is simpler and more correct. PTAL and let me know what you think :)

File: packages/core/src/resource/api.ts [hunk 6]
```diff
@@ -186,11 +186,6 @@ export interface ResourceLoaderParams<R> {
   };
 }
 
-/**
- * Loading function for a `Resource`.
- *
- * @experimental
```
Comment (SkyZeroZx) [inline packages/core/src/resource/api.ts:192]: I don't think we should remove this tsdoc either?

**Comment by surajy93** [comment]
> I have squashed & pushed something that is simpler and more correct. PTAL and let me know what you think :)

yes, it looks good, tested and working fine and sorry next time i will do better and thanks you for your patience and review :)

**Review by surajy93 — COMMENTED** [review]
merge conflict resolved

**Comment by JeanMeche** [comment]
Please rebase instead of merging. We don't want merge commit in our git commit tree.

**Comment by surajy93** [comment]
@JeanMeche 👋 Could you help pass the all checks on this PR when you get a chance?

- The `google-internal-tests` presubmit is failing but I don't have visibility into those logs
- A `target:` label needs to be added

Happy to make any changes needed on my end if the failure is caused by the code changes. Thanks!

**Comment by JeanMeche** [comment]
This is a breaking change and we'll need to fix the usages inside Google's codebase.

File: packages/core/test/resource/params_status_spec.ts [hunk 7]
```diff
@@ -23,7 +23,7 @@ describe('resource with ResourceParamsStatus', () => {
     const s = signal<string | ResourceParamsStatus>('foo');
     const res = await act(() =>
       resource({
-        params: throwStatusAndErrors(s),
+        params: throwStatusAndErrors(s) as () => string,
```
Comment (JeanMeche) [inline packages/core/test/resource/params_status_spec.ts]: This shouldn't be necesary yet it the compiler wasn't happy without the type assertion.

Comment (surajy93) [reply packages/core/test/resource/params_status_spec.ts]: yes looking into it but can you help me to understand what makes google internal test break?

Comment (surajy93) [reply packages/core/test/resource/params_status_spec.ts]: Updated the `params` usage to use an explicit generic instead of a type cast:
params: throwStatusAndErrors<'string'>(s)

Both approaches have identical runtime behavior, but using a generic preserves type safety. 
Unlike `as () => string`, which bypasses type checking, the explicit `<string>` ensures that 
the signal’s value remains assignable to `string | ResourceParamsStatus | Error` and allows 
TypeScript to catch mismatches at compile time.

This change does not alter runtime semantics and maintains compatibility with existing usages.

kindly suggest what do u think?

Comment (surajy93) [reply packages/core/test/resource/params_status_spec.ts]: @JeanMeche can you please let me know if anything need to be done to fixed the google internal tests why it is failing for which reason maybe then i can fixed it.

**Review by surajy93 — COMMENTED** [review]
@JeanMeche made the change and cover all the edge cases and clean the code

**Comment by surajy93** [comment]
@JeanMeche 
please review i have test everything and seems working fine now

Fix: resource / rxResource — Correct params Type When params Is Absent
Branch: fix-core-null-rxresource
Affected APIs: resource(), rxResource()
Files changed:
packages/core/src/resource/resource.ts
packages/core/rxjs-interop/src/rx_resource.ts
packages/core/test/resource/resource_spec.ts
________________________________________
Problem
When calling resource or rxResource with an explicit R type argument but no params
function, TypeScript silently accepted the call and typed the loader/stream params argument
as R. But at runtime, the implementation always falls back to null when no params function
is supplied:

// Implementation fallback (resource.ts):
const params = opts.params ?? oldNameForParams ?? (() => null!);
// always null at runtime when params is omitted
This created a type lie — the loader received null at runtime but TypeScript said it was R.

Broken examples
// Case 1: params omitted entirely with explicit R
resource<string, string>({
  loader: ({params}) => {
    console.log(params); // runtime: null — but TypeScript says: string ❌
    return Promise.resolve('');
  },
});

// Case 2: rxResource with explicit R, no params
rxResource<string, string>({
  stream: ({params}) => {
    console.log(params); // runtime: null — but TypeScript says: string ❌
    return of('');
  },
});
There was also a separate (already-fixed) case where params could be undefined at runtime
because the caller passed a value of type (() => R) | undefined, but this case was covered
by the existing P extends (() => R) | undefined overloads.
________________________________________
Root Cause
The catch-all overloads had params as optional (inherited from BaseResourceOptions), so
TypeScript allowed omitting it even when R was explicitly set to a non-null type. The loader's
params type came from:
// api.ts — ResourceLoaderParams
params: NoInfer<Exclude<R, undefined>>;
With R = string, this evaluates to string — regardless of whether a params function was
actually provided.
________________________________________
Fix
Added a conditional type constraint to both catch-all overloads (with and without defaultValue)
in both resource() and rxResource():
([R] extends [null] ? {} : {params: (ctx: ResourceParamsContext) => R})
Logic:
If R is null (the default) → the constraint is {} → params remains optional → loader gets null ✓
If R is anything else → the constraint requires params → compile error if missing ✓
resource.ts — catch-all overloads (before → after)
// BEFORE
export function resource<T, R = null>(
  options: ResourceOptions<T, R> & {defaultValue: NoInfer<T>},
): ResourceRef<T>;

export function resource<T, R = null>(
  options: ResourceOptions<T, R>,
): ResourceRef<T | undefined>;

// AFTER
export function resource<T, R = null>(
  options: ResourceOptions<T, R> &
    {defaultValue: NoInfer<T>} &
    ([R] extends [null] ? {} : {params: (ctx: ResourceParamsContext) => R}),
): ResourceRef<T>;

export function resource<T, R = null>(
  options: ResourceOptions<T, R> &
    ([R] extends [null] ? {} : {params: (ctx: ResourceParamsContext) => R}),
): ResourceRef<T | undefined>;
The same change was applied symmetrically to rxResource() in rx_resource.ts.
________________________________________
Complete params → loader/stream type matrix (after fix)
params option
R inferred/specified
loader/stream params type
omitted
null (default)
null ✓
undefined (no explicit generics)
null (default)
null ✓
undefined (explicit R given)
e.g. string
string | null ✓
() => R | undefined (may be absent at runtime)
e.g. string
string | null ✓
() => R (always present)
e.g. string
string ✓
omitted but explicit R given
—
compile error ✓
________________________________________
Correct usage examples
// ✅ No params — R defaults to null, loader gets null
resource({
  loader: ({params}) => {
    const _null: null = params; // correctly typed
    return Promise.resolve('hello');
  },
});

// ✅ params: undefined explicitly — loader gets R | null
resource<string, string>({
  params: undefined,
  loader: ({params}) => {
    const _strOrNull: string | null = params; // correctly typed
    return Promise.resolve('hello');
  },
});

// ✅ params may be undefined at runtime — loader gets R | null
function getParams(): (() => string) | undefined {
  return Math.random() > 0.5 ? () => 'foo' : undefined;
}
resource({
  params: getParams(),
  loader: ({params}) => {
    const _strOrNull: string | null = params; // correctly typed
    return Promise.resolve('hello');
  },
});

// ✅ params always present — loader gets R
resource({
  params: () => 'foo',
  loader: ({params}) => {
    const _str: string = params; // correctly typed
    return Promise.resolve('hello');
  },
});

// ❌ Compile error (our fix) — R = string but no params function provided
resource<string, string>({
  loader: ({params}) => Promise.resolve(''), // Error: 'params' is missing
});

// ✅ rxResource — same rules apply, stream must return Observable<T>
rxResource({
  stream: ({params}) => {
    const _null: null = params;
    return of('hello');
  },
});
________________________________________
Test changes
resource_spec.ts — one test was using resource<number | undefined, unknown> without a
params function. The unknown second type arg was unnecessary (the test only exercised
hasValue() behaviour), so it was removed:
// Before
const testResource = resource<number | undefined, unknown>({ loader: ... });

// After
const testResource = resource<number | undefined>({ loader: ... });
All other existing tests — including the full describe('types', ...) block in both
resource_spec.ts and rx_resource_spec.ts — pass without modification.
________________________________________
Verification
packages/core/src/resource/resource.ts — 0 errors ✓
packages/core/rxjs-interop/src/rx_resource.ts — 0 errors ✓
packages/core/test/resource/resource_spec.ts — 0 errors ✓
packages/core/rxjs-interop/test/rx_resource_spec.ts — 0 errors ✓

**Review by surajy93 — COMMENTED** [review]
@JeanMeche please re-run the pipeline

**Review by surajy93 — COMMENTED** [review]
@JeanMeche May I request you to kindly re‑run the Angular trigger? The earlier run seems to have failed.

**Comment by JeanMeche** [comment]
This likely needs a design discussion, there is no perfect answer to the issue (due to some TS limitations). 

Please leave the PR as is for now.

File: packages/core/test/resource/params_status_spec.ts [hunk 8]
```diff
@@ -23,7 +23,7 @@ describe('resource with ResourceParamsStatus', () => {
     const s = signal<string | ResourceParamsStatus>('foo');
     const res = await act(() =>
       resource({
-        params: throwStatusAndErrors(s),
+        params: throwStatusAndErrors<string>(s),
```
Comment (JeanMeche) [inline packages/core/test/resource/params_status_spec.ts:26]: I've illustrated the different workarounds we have for this issue. (see the following changes too)

Comment (surajy93) [reply packages/core/test/resource/params_status_spec.ts:26]: sure thank you :)

**Comment by JeanMeche** [comment]
Closing this stale PR for now
